### DISCUSSION TITLE
Coil_casing and dipoles inside the head using matsimnibs

### DISCUSSION THREAD

**Discussion by martinamancano** [simnibs/simnibs#211] [8. Simulation workflows]
Hi, 

I'm running TMS simulations using the matsimnibs matrix imported from Softaxic .stmpx files, and the coil file MagVenture_MCF-B65.ccd, corresponding to the coil model MCF-B65 I used collecting data. 

However, the results of the stimulation are weird, in that the coil is represented in the correct position and handle orientation, but it looks like it overlays partially with the scalp, in every simulation I run. 

Does it mean there's a problem in the coordinates of the coil?

Thank you in advance, 
Martina

**Comment by TorgeW** [comment]
What do you mean with every simulation? Is that for multiple subjects in multiple experiments or just one subject?

Could you also share the .stmpx file and maybe add some screenshots where we can see the intersection?

**Reply by martinamancano** [reply]
I'm running a for loop for all the subjects included in the experiment, with a nested loop for all the coordinates they were stimulated from (I had multiple TMS sessions for each subject).

That's an example of my .stmpx file:
[Subj_03_RI_RS - Copy.txt](https://github.com/simnibs/simnibs/files/12870784/Subj_03_RI_RS.-.Copy.txt) (I converted it to .txt to share it here)

and these are the corresponding numpy matrices where I saved the coordinates (attaching the screeshots):
![image](https://github.com/simnibs/simnibs/assets/130748492/41c3735e-d45f-4ca4-92ed-5665bf15d4f5)
![image](https://github.com/simnibs/simnibs/assets/130748492/d8b504fe-04d5-400c-87d4-f3f697ce751b)
![image](https://github.com/simnibs/simnibs/assets/130748492/1d530ec8-8fd5-4c35-a614-8ecf8f708bd8)


and that's the part of the code I used to extract the matsimnibs matrices:
`
            
            # Initialize session       
            s = sim_struct.SESSION()
            fn = item.name # filename of softaxic points
            
            fn_parts = fn.split("_") # separate
            Subj_ID = fn_parts[0] + "_" + fn_parts[1] # get subj ID from the filename
            
            tms_list = softaxic().read(fn)  # read all targets from file and return as TMSLIST()
            s.add_tmslist(tms_list)
            
            for point in range(len(tms_list.pos)): # for every point in that .stmpx file
                p_name = tms_list.pos[point].name  # <- name is filled with pos-ID from .stmpx.
                p_mat = tms_list.pos[point].matsimnibs # <- matrix of coordinates of the coil
                
                subj_p_name = Subj_ID + '_' + p_name # concatenate with point name
                print(subj_p_name)
                
                # Define the subfolder path using pathlib
                subfolder_path = Path("matsimnibs coo")
                
                # Define the full path to save the array within the subfolder
                file_path = subfolder_path / subj_p_name
                
                # Save the matsimnibs array with coil position and orientation info
                np.save(file_path, p_mat)`

then I'm using another script to run the simulation, that's the part where I'm setting each simulation parameters:

`              

                 ### Initialize a session
                S = sim_struct.SESSION()
                S.subpath = head.name  # m2m-folder of the subject 
                
                # Don't open in gmsh
                S.open_in_gmsh = False
           
                # Define general subfolder path
                subfolder_path = Path("TMS Simulations")
                
                # define simulation name
                sim_name = coo.name.replace(".npy", "")
                print(sim_name)
                # sub-subfolder of specific sessions
                sim_name_parts = sim_name.split("_")
                session = Path(sim_name_parts[-1]) 
                
                # define simulation path
                sim_path = subfolder_path / session / sim_name
                
                S.pathfem = sim_path  # Directory where the simulation will be saved
                
                ## Define the TMS simulation
                tms = S.add_tmslist()
                tms.fnamecoil = os.path.join('coil_ID','MagVenture_MCF-B65.ccd')  # Choose a coil model
                
                # load simnibs coordinates .npy file
                my_pos = np.load(coo)
                
                # Define the coil position
                pos = tms.add_position()
                pos.matsimnibs = my_pos # specify coil centre and orientation
                
                # define the stimulation intensity
                pos.didt = dIdt_Ams * 1000 # to convert from A/ms to A/s
                print(pos.didt)
                
                # Run Simulation
                run_simnibs(S)`

screenshots of the results I'm getting:
![image](https://github.com/simnibs/simnibs/assets/130748492/6ebbe245-d697-4d31-b134-b5c6e7f2658d)
![image](https://github.com/simnibs/simnibs/assets/130748492/37782649-ba51-448f-995d-0ac43d55a397)

As you can see, the coil case overlaps with both scalp and brain, while it should be represented over the head surface.

**Reply by martinamancano** [reply]
1) 
![image](https://github.com/simnibs/simnibs/assets/130748492/d4a9a70f-dfb0-4ef4-9a62-817171040671)

I used the same T1, but I also added the -forceqform to run the head mesh. Might this be the issue?

2) I tried to change approach, extracting the coil centre coordinates from the matsimnibs matrix, following this:
![image](https://github.com/simnibs/simnibs/assets/130748492/9f0eab0e-6484-4321-97e5-d0bd66da2a9f)
However, when I tried to simply pick up the 3 y coordinates, it didn't work and the coil ended up in the wrong orientation. How do I extract the handle orientation from matsimnibs?
Here's what I did:
`coil_centre = my_pos[(0, 1, 2), 3]
handle_dir = my_pos[(0, 1, 2), 1]`

**Comment by oulap** [comment]
To me it looks like the target position is on the cortex and that's why the coil ends up in the head. The py-column of the matrix defines the  direction pointing away from the handle, see here:

https://simnibs.github.io/simnibs/build/html/_downloads/46a4ff127db96296fb4238660aec96f0/coil_axesorientation.pdf

I'm not sure why you would want to use that as your coil center (the last column). If you are trying to project the coil to the scalp you can try to use this function:

https://github.com/simnibs/simnibs/blob/fbc36c0ead76fa1cca057ae3e93f388efac31f75/simnibs/utils/transformations.py#L777

which takes in the head mesh, the points you want to project (the target position/last column of matsimnibs), and a surface tag (1005 for scalp).

**Reply by martinamancano** [reply]
1) The target position is not on the cortex in Softaxic space or in T1 space either:  if I insert the three coordinates in MRIcroGL, they lie on the head surface where the coil was touching the head (and not on the cortex as it looks in the .msh file from the simulation), e.g.:
![image](https://github.com/simnibs/simnibs/assets/130748492/683a6c25-89b4-4a70-aca9-0e8ab2db00cc)

and if I run a simulation with pos.centre = my_pos[(0, 1, 2), 3] and pos.distance = 1, the coil ends up over the head, not within it, e.g. (different simulation):
![image](https://github.com/simnibs/simnibs/assets/130748492/1d1f71a2-d30c-4ab8-a96a-99cd2eb517b9)

Softaxic also saves the corresponding point on the brain, but that's not what SimNIBS import, because the function to read the Softaxic coordinates (softaxic().read(fn) https://github.com/simnibs/simnibs/blob/fbc36c0ead76fa1cca057ae3e93f388efac31f75/simnibs/utils/nnav.py#L4) imports the fp attributes, where position of the coil on the head and orientation are saved, from what I understand. The documentation says: "pc is the center of the coil", and they corresponds to the coordinates appearing in the Softaxic session over the head, that's why I'm picking them for the pos.centre attribute.
However, the coil does end up inside the head in the simulation, and I don't understand why.

2) I understand now the orientation of the axis on SimNIBS, and I guess Softaxic has a different one. I was trying to define the direction pointing _towards_ the handle (or, where the handle was positioned). Is there a way to obtain that, using pos_ydir?

3) Thank you for the suggestion, I'll try to use it! [](url)

**Reply by oulap** [reply]
2. The import should have taken care of transforming the coordinate system from softaxic to the simnibs one, see here:

https://github.com/simnibs/simnibs/blob/fbc36c0ead76fa1cca057ae3e93f388efac31f75/simnibs/utils/nnav.py#L358

but maybe something goes wrong there. I can try to have a look at the smptx file you sent. 

py points away from the handle in the simnibs coil coordinate system so if you want the direction of the handle you can just multiply the direction by -1.

**Reply by martinamancano** [reply]
I don't know how to use the _project_points_on_surface_ function. Here's my code:
`
#import simnibs
import os
from simnibs import transformations
import numpy as np`

`#change directory`
`os.chdir("E:\My Drive\TMS experiment\MR subjs\T1 numbered\m2m_Subj_05.nii")`

`my_pos = np.load("E:\My Drive\TMS experiment\Punti Softaxic\only COIL coo\matsimnibs coo\Subj_05_I001_RS.npy")`

`centre = my_pos[(0, 1, 2), 3]`

`transformed = transformations.project_points_on_surface(mesh = "m2m_Subj_05.nii.msh", pts = centre, surface_tags=1005)`

and it gives me this error:

`runcell(0, 'E:/My Drive/TMS experiment/MR subjs/script di prova/project.py')
Traceback (most recent call last):`

  `File "C:\Users\martina.mancano\SimNIBS-4.0\simnibs_env\lib\site-packages\spyder_kernels\py3compat.py", line 356, in compat_exec
    exec(code, globals, locals)`

  `File "e:\my drive\tms experiment\mr subjs\script di prova\project.py", line 22, in <module>
    transformed = transformations.project_points_on_surface(mesh = "m2m_Subj_05.nii.msh", pts = centre, surface_tags=1005)`

   `File "C:\Users\martina.mancano\SimNIBS-4.0\simnibs_env\lib\site-packages\simnibs\utils\transformations.py", line 808, in project_points_on_surface
    tr_of_interest = (mesh.elm.elm_type == 2) * (np.in1d(mesh.elm.tag1, surface_tags)) `

`AttributeError: 'str' object has no attribute 'elm'`

2) I tried to multiply the ydir by -1, but it doesn't work, the coil ends in the same, wrong direction regardless of the multiplication:
my_pos = np.load("name of the numpy array with matsimnibs matrix")

coil_centre = my_pos[(0, 1, 2), 3]
handle_dir = my_pos[(0, 1, 2), 1]
handle_dir = handle_dir*-1 

It seems like the handle_dir attribute is not affecting the simulation if I use numerical coordinates, whereas if I use electrode position (e.g. 'F3') it responds.

**Comment by oulap** [comment]
The projection function expects a mesh object, so you need to load it first using something like

```
from simnibs.mesh_tools import mesh_io
m = mesh_io.read_msh('my_mesh.msh')
```
and then pass the variable m to the projection function.

The simulation either uses matsimnibs or the position (and then constructs the matrix from there). You should be able to prepare the simulation (in a script) and then substitute your custom matsimnibs into the poslist just before calling the run_simnibs-function. Specifically the poslist object has a field called matsimnibs, which you can overwrite using the custom matrix you have.

**Reply by martinamancano** [reply]
1) ok, thank you!

2) Yeah, indeed at the beginning I tried using pos.matsimnibs. So if I use:
a) **pos.matsimnibs**:  coil in the right position and handle orientation, but shifted inside the head
b) **pos.centre, pos_ydir and pos.distance**: coil in the right centre, and I can decide the right distance from the head, but I cannot rotate the handle direction (multiplying by -1 for some mysterious reason is not affecting the handle orientation). 

So with either way, there's something that does not work .

3) I uninstalled SimNIBS, because when I tried to create a new head mesh but it wouldn't work, and it looked like the error was related to the numpy version (I think when I installed pandas in simnibs, it also changed the numpy version). Anyway, I tried to re-install SimNIBS 4.0.0 but it's not working. That's the final message, at the end on the installation: `ImportError: Trying to run post-install script without PyQt istalled, please use the silent mode (postinstall_simnibs --help for more information
PyQt workaround failed`

I also attach the log file:
[install.log](https://github.com/simnibs/simnibs/files/13052556/install.log)

I also tried with SimNIBS 4.0.1, but then I have another message of error when I try to install Spyder.

**Reply by martinamancano** [reply]
Hello, 

- I resorted to install SimNIBS in silent mode with Anaconda, but this way I cannot use the GUI (I think it's related to QyPt5 and PyQt, even though I installed them, for some reason it's not able to import them), and it gives me this error when I use charm: 

`C:\Users\martina.mancano\Miniconda3\envs\simnibs_env\lib\site-packages\requests\__init__.py:102: RequestsDependencyWarning: urllib3 (1.26.18) or chardet (5.2.0)/charset_normalizer (2.0.4) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({})/charset_normalizer ({}) doesn't match a supported "
Traceback (most recent call last):
  File "C:\Users\martina.mancano\Miniconda3\envs\simnibs_env\Lib\site-packages\simnibs\cli\charm.py", line 31, in <module>
    from simnibs.segmentation import charm_main
  File "C:\Users\martina.mancano\Miniconda3\envs\simnibs_env\lib\site-packages\simnibs\segmentation\charm_main.py", line 15, in <module>
    from . import samseg
  File "C:\Users\martina.mancano\Miniconda3\envs\simnibs_env\lib\site-packages\simnibs\segmentation\samseg\__init__.py", line 10, in <module>
    from .AffineWholeHead import AffineWholeHead
  File "C:\Users\martina.mancano\Miniconda3\envs\simnibs_env\lib\site-packages\simnibs\segmentation\samseg\AffineWholeHead.py", line 18, in <module>
    import matplotlib.pyplot as plt
ModuleNotFoundError: No module named 'matplotlib'` 
It never gave me this error when I installaed SimNIBS the first time. 

- I'm doing the group analysis on MATLAB, and I would like to compute the average electric field in ROI in MNI space. 
I found [here](https://simnibs.github.io/simnibs/build/html/tutorial/analysis.html) this tutorial that explains how to do them at the single-subject level, and [here](https://simnibs.github.io/simnibs/build/html/tutorial/advanced/group_analysis.html) the one about group analysis, but there's only the part with fsaverage and ROI, and nothing with MNI. Is there a way to do it? I already calculated the mean electric field in MNI space in nifti volume, could I use that file?
Moreover, when I follow the steps with Fsaverage and try to load the the atlas, it gives me this error:
`[m, snames]=mesh_load_fssurf('fsaverage','label','HCP_MMP1');`
`Unrecognized function or variable 'gifti'.

Error in mesh_load_fssurf>load_surface (line 162)
        s=gifti(fname);

Error in mesh_load_fssurf (line 117)
    [nodes, triangles] = load_surface(lh{1});`

**Reply by oulap** [reply]
Try installing matplotlib in the python environment to get rid of the charm error.

For MNI analysis you can use the "map_to_mni" flag when running the simulations, see here:

https://simnibs.github.io/simnibs/build/html/documentation/sim_struct/session.html

This will map the electric field results to MNI space for each subject simulation. You can then use any atlas defined in MNI space to analyze the results within the atlas regions you want.

When I call:

`[m, snames]=mesh_load_fssurf('fsaverage','label','HCP_MMP1');`

that seems to work fine. Could you check that the path where the files are being read from is correct? You can just edit or set a breakpoint in the `mesh_load_fssurf` matlab function to inspect the file path that is being passed to the reader.

**Comment by cascam07** [comment]
@oulap @martinamancano I have been encountering this same issue but using coordinates exported from BrainSight. Coil shows up in the right location and orientation, but it is partially overlapping with the scalp. 

![image](https://github.com/simnibs/simnibs/assets/5588390/df10fd1e-7c11-4fa3-98d9-b02151e6ba3d)

What I suspect is happening is that the definition of what the coordinate represents in the neuronavigation system and in SimNIBS are not the same. For example, if Softaxic/Brainsight are tracking the coil coordinates as the center of the bottom surface of the coil, i.e. where the coil is making contact with the scalp, but SimNIBS is treating the coordinate as the center of the coil volume, i.e. inside of the coil, that would result in the coil always showing up half a coil's width *into* the head. Is this consistent with how SimNIBS interprets coil coordinates? Is it the center of the coil volume or the center of the inner surface?

**Reply by TorgeW** [reply]
SimNibs is using the center of the bottom surface of the coil, i.e. where the coil is making contact with the scalp for most coils. 
Did you use the same T1 for the navigation and SimNIBS?

**Reply by cascam07** [reply]
Yes, I used the same T1 for SimNIBS and neuronavigation. However, similar to martinamancano, I did use the --forceqform option to get charm to run. Because of our study design, there is not time to process the scans *before* the TMS, so the same T1 was used for both SimNIBS and neuronavigation, the T1 that SimNIBS output was not used for neuronavigation.

**Reply by cascam07** [reply]
I just noticed that you said "for most coils". Is there a good way to check where on the coil SimNIBS is using as coil center? Is it just the origin of the coil file?

**Reply by TorgeW** [reply]
Yes it is the origin of the coil file. 
For most coils includes all fig of 8 coils but excludes some of the more special coils.

**Reply by cascam07** [reply]
@TorgeW I figured out what was happening in my case at least. I realized that the coil is not in the same orientation in the Brainsight coordinates vs SimNIBS coordinates (X and Z are flipped). The Python interface has a brainsight().read() function that incorporates the necessary transform. I've been running my simulations through Matlab, however, which does not have a similar read function (as far as I can see). Once I incorporated the coordinate transform into my Matlab code, the coil is showing up in the correct position on the scalp.

**Reply by BPassera** [reply]
Hi @cascam07 

How did you implement this transformation ? I have been having the same issue.

**Reply by TorgeW** [reply]
@BPassera 
Please use brainsight().read() as @cascam07 says this fixed his problem.

**Reply by cascam07** [reply]
@BPassera @TorgeW 
It's true that brainsight().read() will solve the problem, but that's only available through the Python interface. If you are using Matlab to run your simulations, there isn't a built in function you can use (as far as I know).

Here is how I applied the same transform, specified in brainsight().read(), in Matlab.

```
log_table = readtable('../Brainsight_Log_File.txt'); %Your export from Brainsight
ses = sim_struct('SESSION');
ses.poslist{1} = sim_struct('TMSLIST');

bs_trsform = [-1, 0, 0, 0;  % +x -> -x
             0, 1, 0, 0;    % +y -> +y
             0, 0, -1, 0;   % +z -> -z
             0, 0, 0, 1];

for t_idx = 1:height(log_table)
    T = log_table(t_idx,:);
    coordmat  = [T.m0n0, T.m1n0, T.m2n0, T.Loc_X;
                T.m0n1, T.m1n1, T.m2n1, T.Loc_Y;
                T.m0n2, T.m1n2, T.m2n2, T.Loc_Z;
                     0,      0,      0,       1];
    ses.poslist{1}.pos(t_idx).matsimnibs = coordmat * bs_trsform;
end
```
